src/cbs.c

```diff
@@ -256,54 +256,79 @@ static void on_amqp_management_execute_operation_complete(void* context, AMQP_MA
-                free(result);
-                result = NULL;
             }
             else
             {
                 /* Codes_SRS_CBS_01_034: [ `cbs_create` shall create an AMQP management handle by calling `amqp_management_create`. ]*/
                 /* Codes_SRS_CBS_01_002: [ Tokens are communicated between AMQP peers by sending specially-formatted AMQP messages to the Claims-based Security Node. ]*/
                 /* Codes_SRS_CBS_01_003: [ The mechanism follows the scheme defined in the AMQP Management specification [AMQPMAN]. ]*/
-                result->amqp_management = amqp_management_create(session, "$cbs");
-                if (result->amqp_management == NULL)
+                cbs->amqp_management = amqp_management_create(session, "$cbs");
+                if (cbs->amqp_management == NULL)
                 {
-                    free(result);
-                    result = NULL;
+                    LogError("Cannot create AMQP management instance for the $cbs node.");
                 }
                 else
                 {
-                    result->cbs_state = CBS_STATE_CLOSED;
+                    /* Codes_SRS_CBS_01_116: [ If setting the override key names fails, then `cbs_create` shall fail and return NULL. ]*/
+                    if (amqp_management_set_override_status_code_key_name(cbs->amqp_management, "status-code") != 0)
+                    {
+                        /* Codes_SRS_CBS_01_116: [ If setting the override key names fails, then `cbs_create` shall fail and return NULL. ]*/
+                        LogError("Cannot set the override status code key name");
+                    }
+                    else
+                    {
+                        /* Codes_SRS_CBS_01_118: [ `cbs_create` shall set the override status description key name on the AMQP management handle to `status-description` by calling `amqp_management_set_override_status_description_key_name`. ]*/
+                        if (amqp_management_set_override_status_description_key_name(cbs->amqp_management, "status-description") != 0)
+                        {
+                            /* Codes_SRS_CBS_01_116: [ If setting the override key names fails, then `cbs_create` shall fail and return NULL. ]*/
+                            LogError("Cannot set the override status description key name");
+                        }
+                        else
+                        {
+                            /* Codes_SRS_CBS_01_001: [ `cbs_create` shall create a new CBS instance and on success return a non-NULL handle to it. ]*/
+                            cbs->cbs_state = CBS_STATE_CLOSED;
+
+                            goto all_ok;
+                        }
+                    }
+
+                    amqp_management_destroy(cbs->amqp_management);
                 }
+
+                singlylinkedlist_destroy(cbs->pending_operations);
             }
+
+            free(cbs);
```

Comment:
>free(cbs); [](start = 12, length = 10)

Nit - instead of individually freeing up members above in error case, could you just invoke cbs_destroy() and let it hide all the details?

Reply:
I prefer writing it this way. The reason is that it is easier for me to track the frees in the same method. And if I were to use the call to destroy for cleanup of create then I'd have to make sure that every member gets initialized with NULL prior to executing the code.

Not saying it cannot be done, but the result is pretty much the same (and tests have to check both ways of writing it it anyhow), thus changing it would be 0 ROI.

---
In reply to: [167072243](https://github.com/Azure/azure-uamqp-c/pull/211#discussion_r167072243) [](ancestors = 167072243)